### PR TITLE
Remove use of jQuery events from cockpit.js

### PR DESCRIPTION
**Pull request by stefwalter**
In order to future proof cockpit.js we should remove jQuery. This
is the first of a few commits to that end.

### CODE REVIEW

**Comment by stefwalter** [comment]
Fixed issue with DBusProxies

File: pkg/base1/cockpit.js [hunk 1]
```diff
@@ -153,30 +153,54 @@ function transport_debug() {
...
+            value: function addEventListener(type, handler) {
+                if (handlers[type] === undefined)
+                    handlers[type] = [ ];
+                handlers[type].push(handler);
+            }
+        },
+        removeEventListener: {
+            enumerable: false,
+            value: function removeEventListener(type, handler) {
+                var length = handlers[type] ? handlers[type].length : 0;
```
Comment (dperpeet) [inline pkg/base1/cockpit.js]: since we don't have to worry about concurrency, why not use something like this here:
```
var index = handlers[type] ? handlers[type].indexOf(handler) : -1;
if (index > -1)
    handlers[type].splice(index, 1);
```
That seems clearer and we don't have null values cluttering the array.

File: pkg/base1/cockpit.js [hunk 2]
```diff
@@ -153,30 +153,54 @@ function transport_debug() {
...
+                for (var i = 0; i < length; i++) {
+                    if (handlers[type][i] === handler) {
+                        handlers[type][i] = null;
+                        break;
+                    }
+                }
+            }
+        },
+        dispatchEvent: {
```
Comment (dperpeet) [inline pkg/base1/cockpit.js]: I think a comment would be useful here for the special "string" parameter mode.

File: pkg/base1/cockpit.js [hunk 3]
```diff
@@ -1606,9 +1630,7 @@ function basic_scope(cockpit, jquery) {
                 callback.call(self, row, x, n);
             }
 
-            var event = document.createEvent("CustomEvent");
-            event.initCustomEvent("notify", false, false, [ x, n ]);
-            self.dispatchEvent(event, x, n);
+            self.dispatchEvent("notify", x, n);
```
Comment (dperpeet) [inline pkg/base1/cockpit.js]: I'm not sure, but this seems like it changes the behavior:
https://github.com/cockpit-project/cockpit/pull/4145/files#diff-34cbdb779d20a4a75591fcba50bd25d4R183

File: pkg/base1/cockpit.js [hunk 4]
```diff
@@ -2348,16 +2375,19 @@ function full_scope(cockpit, $, po) {
...
 
         function signal(path, iface, name, args) {
-            $(self).triggerHandler("signal", [name, args]);
-            if (name[0].toLowerCase() != name[0])
-                $(self).triggerHandler(name, args);
+            self.dispatchEvent("signal", name, args);
+            if (name[0].toLowerCase() != name[0]) {
+                args = args.slice();
+                args.unshift(name);
+                self.dispatchEvent.apply(self, args);
```
Comment (dperpeet) [inline pkg/base1/cockpit.js]: Are you certain about this use of apply directly on self.dispatchEvent?

File: pkg/base1/cockpit.js [hunk 5]
```diff
@@ -153,30 +153,54 @@ function transport_debug() {
 }
 
 function event_mixin(obj, handlers) {
```
Comment (dperpeet) [inline pkg/base1/cockpit.js]: This function would benefit from a brief description.

File: pkg/base1/cockpit.js [hunk 6]
```diff
@@ -3461,85 +3502,85 @@ function full_scope(cockpit, $, po) {
...
+                    timestamp = 0;
+                    if (meta.now && meta.timestamp)
+                        timestamp = meta.timestamp + ($.now() - meta.now);
+                    beg = Math.floor(timestamp / interval);
+                    callback(beg, meta, null, options_list[0]);
+
+                /* A data message */
+                } else if (meta) {
+
+                    /* Data decompression */
```
Comment (dperpeet) [inline pkg/base1/cockpit.js]: I know this is from previous code, but is there a bit more to write here? Either a description of what kind of decompression or even a link to the corresponding compression would be nice.

File: pkg/base1/test-metrics.html [hunk 7]
```diff
@@ -77,14 +77,55 @@ <h2 id="qunit-userAgent"></h2><ol id="qunit-tests"></ol>
...
             channel.valid = false;
-            $(channel).trigger("close", [options || { }]);
+            channel.dispatchEvent("close", options || { });
         }
     }
 
     var peer = this;
     var last_channel = 0;
 
+    function event_mixin(obj, handlers) {
```
Comment (dperpeet) [inline pkg/base1/test-metrics.html]: Is there a better way than copy/pasting this? Or at least note the source?

**Comment by dperpeet** [comment]
Finished reviewing. Good changes, only a few comments.

File: pkg/base1/cockpit.js (hunk 5 above)
Comment (stefwalter) [inline pkg/base1/cockpit.js]: Added.

File: pkg/base1/cockpit.js (hunk 1 above)
Comment (stefwalter) [inline pkg/base1/cockpit.js]: We're not worried about concurrency but reentrancy. The code above will break the latter.

File: pkg/base1/cockpit.js (hunk 2 above)
Comment (stefwalter) [inline pkg/base1/cockpit.js]: Added into the top comment.

File: pkg/base1/cockpit.js (hunk 3 above)
Comment (stefwalter) [inline pkg/base1/cockpit.js]: We don't use the details argument so far. But I fixed this anyway so it works for the general case where someone does use the details argument.

File: pkg/base1/cockpit.js (hunk 4 above)
Comment (stefwalter) [inline pkg/base1/cockpit.js]: Yes. Why not?

File: pkg/base1/cockpit.js (hunk 6 above)
Comment (stefwalter) [inline pkg/base1/cockpit.js]: Sure. But that's not part of this pull request. An additional pull request would be welcome.

File: pkg/base1/test-metrics.html (hunk 7 above)
Comment (stefwalter) [inline pkg/base1/test-metrics.html]: Fixed.

**Comment by stefwalter** [comment]
Updated after review.

File: pkg/base1/cockpit.js (hunk 1 above)
Comment (dperpeet) [inline pkg/base1/cockpit.js]: > We're not worried about concurrency but reentrancy. The code above will break the latter.

Why? Where do we store an index to an entry in the array for later use? Other than that I don't see how the javascript execution model would allow problems to arise from shrinking the array.

Comment (dperpeet) [inline pkg/base1/cockpit.js]: I see the point you made about arbitrary code in `dispatchEvent`.

**Comment by dperpeet** [comment]
Ready to merge once tests pass.
